src/vuln_analysis/utils/functions_parsers/java_functions_parsers.py

```diff
@@ -0,0 +1,2662 @@
+                    elif ch == '\\': esc = True
+                    elif ch == "'": in_chr = False
+                    i += 1; continue
+                if ch == '"': in_str = True; i += 1; continue
+                if ch == "'": in_chr = True; i += 1; continue
+                if ch == '(':
+                    depth += 1
+                elif ch == ')':
+                    depth -= 1
+                    if depth == 0:
+                        return i
+                i += 1
+            return -1
+
+        def _next_token_after(s: str, idx: int) -> str:
+            """Return the next non-space token after idx: '{', 'throws', or first char/empty."""
+            n = len(s)
+            i = idx + 1
+            while i < n and s[i].isspace():
+                i += 1
+            if s.startswith('throws', i) and (i + 6 == n or not s[i + 6].isalnum() and s[i + 6] != '_'):
+                return 'throws'
+            return s[i] if i < n else ''
+
+        # New: precise left-scan to the start of the expression that owns this call.
+        # Stops at the first top-level boundary (not inside (),[],{} or strings):
+        # ; , = + - * / % ! ~ ? : & | ^ < > { } \n
+        # We allow dots, identifiers, casts, and 'new ...' as part of the expression.
+        BOUNDARY_CHARS = set(';,=+-*/%!?::&|^<>{}\n')
+        def _expr_start_left(s: str, pos: int, max_back: int = 512) -> int:
+            i = pos - 1
+            in_str = in_chr = False
+            esc = False
+            # depths for (), [], {}
+            dp = db = dbr = 0
+            limit = max(0, pos - max_back)
+            while i >= limit:
+                ch = s[i]
+                if in_str:
+                    if esc: esc = False
+                    elif ch == '\\': esc = True
+                    elif ch == '"': in_str = False
+                    i -= 1; continue
+                if in_chr:
+                    if esc: esc = False
+                    elif ch == '\\': esc = True
+                    elif ch == "'": in_chr = False
+                    i -= 1; continue
+                if ch == '"': in_str = True; i -= 1; continue
+                if ch == "'": in_chr = True; i -= 1; continue
+
+                # Track bracket depths (we treat scanning left; depths just indicate "inside something")
+                if ch == ')': dp += 1; i -= 1; continue
+                if ch == '(': dp = max(0, dp - 1); i -= 1; continue
+                if ch == ']': db += 1; i -= 1; continue
+                if ch == '[': db = max(0, db - 1); i -= 1; continue
+                if ch == '}': dbr += 1; i -= 1; continue
+                if ch == '{': dbr = max(0, dbr - 1); i -= 1; continue
+
+                # Only consider boundaries at top level
+                if dp == db == dbr == 0 and ch in BOUNDARY_CHARS:
+                    return i + 1  # start right after the boundary
+
+                i -= 1
+            return limit  # we hit the scan cap; good enough
+
```

Comment:
@tmihalac These 3 closure functions are splitting the logic of the `search_for_called_function`, thus it hurts the clarity and the readability of the code.
IMO, it's better placing all these 3 closure functions as block at the beginning of the `search_for_called_function` , so it will not break the logic continuation ( logic of `search_for_called_function` itself will follow after the block of 3 closure functions).

Reply:
It's possible if the closures are not accessing the outer scope' variables/references.